src/viam/examples/modules/complex/base/impl.cpp

```diff
@@ -0,0 +1,96 @@
+    auto base_name = cfg.name();
+    auto motor = cfg.attributes()->find(motor_name);
+    if (motor == cfg.attributes()->end()) {
+        std::ostringstream buffer;
+        buffer << base_name << ": Required parameter `" << motor_name
+               << "` not found in configuration";
+        throw std::invalid_argument(buffer.str());
+    }
+    const auto* const motor_string = motor->second->get<std::string>();
+    if (!motor_string || motor_string->empty()) {
+        std::ostringstream buffer;
+        buffer << base_name << ": Required non-empty string parameter `" << motor_name
+               << "` is either not a string "
+                  "or is an empty string";
+        throw std::invalid_argument(buffer.str());
+    }
+    return *motor_string;
+}
+
+void MyBase::reconfigure(Dependencies deps, ResourceConfig cfg) {
+    // Downcast `left` and `right` dependencies to motors.
+    auto left = find_motor(cfg, "left");
+    auto right = find_motor(cfg, "right");
+    for (const auto& kv : deps) {
+        if (kv.first.short_name() == left) {
+            left_ = std::dynamic_pointer_cast<Motor>(kv.second);
+        }
+        if (kv.first.short_name() == right) {
+            right_ = std::dynamic_pointer_cast<Motor>(kv.second);
+        }
+    }
+}
+
+std::vector<std::string> MyBase::validate(ResourceConfig cfg) {
+    return {find_motor(cfg, "left"), find_motor(cfg, "right")};
+}
+
+bool MyBase::is_moving() {
+    return left_->is_moving() || right_->is_moving();
+}
+
+grpc::StatusCode MyBase::stop(const AttributeMap& extra) {
+    auto left_stop = left_->stop(extra);
+    if (left_stop != grpc::StatusCode::OK) {
+        return left_stop;
+    }
+    auto right_stop = right_->stop(extra);
+    if (right_stop != grpc::StatusCode::OK) {
+        return right_stop;
+    }
+    return grpc::StatusCode::OK;
+}
+
+void MyBase::set_power(const Vector3& linear, const Vector3& angular, const AttributeMap& extra) {
+    // Stop the base if absolute value of linear and angular velocity is less
+    // than 0.01.
+    if (abs(linear.y()) < 0.01 && abs(angular.z()) < 0.01) {
+        stop(extra);  // ignore returned status code from stop
+        return;
+    }
+
+    // Use linear and angular velocity to calculate percentage of max power to
+    // pass to set_power for left & right motors
+    auto sum = abs(linear.y()) + abs(angular.z());
+    left_->set_power(((linear.y() - angular.z()) / sum), extra);
+    right_->set_power(((linear.y() + angular.z()) / sum), extra);
+}
+
+AttributeMap MyBase::do_command(const AttributeMap& command) {
+    std::cout << "Received DoCommand request for MyBase " << Resource::name() << std::endl;
+    return command;
+}
+
+std::vector<GeometryConfig> MyBase::get_geometries(const AttributeMap& extra) {
+    auto left_geometries = left_->get_geometries(extra);
+    auto right_geometries = right_->get_geometries(extra);
+    std::vector<GeometryConfig> geometries(left_geometries);
```

Comment:
Instead of creating the vector like this and doing the insert I think you could probably just do `return {left_geometries, right_geometries}`

Reply:
I tried that originally, but there's apparently no overload of vector constructor that allows that. Stole from this SO post https://stackoverflow.com/questions/3177241/what-is-the-best-way-to-concatenate-two-vectors, which is admittedly a bit old.

Reply:
I think this will become possible with C++20 / C++23 `range` support, but that's only of academic interest for us.